ob1k-security/src/main/java/com/outbrain/ob1k/security/server/PathAssociations.java

```diff
@@ -0,0 +1,164 @@
+package com.outbrain.ob1k.security.server;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Created by gmarom on 6/9/15
+ * <p/>
+ * <p>
+ * Provides services related with path->authenticator associations.
+ * </p>
+ * <p>
+ * This class cannot be constructed directly, only through the {@code PathAssociationsBuilder} class.
+ * </p>
+ *
+ * @param <T> the type of the class used in the authenticators associated with the paths,
+ *           such as the {@code UserPasswordToken}
+ */
+public class PathAssociations<T> {
+
+  public static final String PATH_DELIMITER = "/";
+  //Key: An endpoint relative path
+  //Value: An instance of a CredentialsAuthenticator
+  //This is a multi map, so a URL might have several CredentialAuthenticators associated with it
+  private final SetMultimap<String, CredentialsAuthenticator<T>> associations = HashMultimap.create();
+
+  private PathAssociations(final SetMultimap<String, CredentialsAuthenticator<T>> associations) {
+    this.associations.putAll(associations);
+  }
+
+  /**
+   * @return all the {@code CredentialAuthenticators} associated with the given {@code path}
+   * and an empty set if none exist
+   */
+  public Set<CredentialsAuthenticator<T>> getAuthenticators(final String path) {
+    final Set<CredentialsAuthenticator<T>> authenticators = new HashSet<>(associations.size());
+    return accumulateAuthenticators(validatePath(path), authenticators);
+  }
+
+  private Set<CredentialsAuthenticator<T>> accumulateAuthenticators(final String path,
+                                                                    final Set<CredentialsAuthenticator<T>> authenticators) {
+    authenticators.addAll(associations.get(path));
+
+    if (isRootPath(path)) return authenticators;
+    else return accumulateAuthenticators(parentPath(path), authenticators);
+  }
+
+  /**
+   * <p>
+   * Checks whether or not {@code authenticator} may authorize access to {@code path}. Such an authenticator exists
+   * if it was associated with {@code path} or with any of its parents. For example for the path "/a/b/c" Any authenticator
+   * assigned with either "/a/b/c", "/a/b", "/a", "/" is valid
+   * </p>
+   * <p/>
+   * <p>
+   * <table>
+   * <tr>
+   * <th>Authenticator associated with</th>
+   * <th>Authorized paths</th>
+   * <th>Unauthorized paths</th>
+   * </tr>
+   * <tr>
+   * <td>/</td>
+   * <td>/*</td>
+   * <td>N/A</td>
+   * </tr>
+   * <tr>
+   * <td>/service1</td>
+   * <td>/service1/endpoint1<br>/service1/sub1/endpoint2</td>
+   * <td>/service2/endpoint2</td>
+   * </tr>
+   * </table>
+   * </p>
+   */
+  public boolean isAuthorized(final CredentialsAuthenticator authenticator, final String path) {
```

Comment:
doesn't make sense. this class handles path mappings. it should know nothing about authorization